solutions/CF Orac-LCM.mdx

```diff
@@ -25,22 +28,67 @@ ll a,b;
 int n;
 
 void red() { // restore invariant mentioned above
-	ll g = __gcd(a,b);
-	tie(a,b) = make_pair(g,a/g*b);
+    ll g = __gcd(a,b);
+    tie(a,b) = make_pair(g,a/g*b);
 }
 
 int main() {
-	cin >> n >> a >> b;
-	red();
-	for (int i = 2; i < n; i++) {
-		ll x; cin >> x; 
-		b = __gcd(b,x);
-		red();
-	}
-	cout << b << endl;
+    cin >> n >> a >> b;
+    red();
+    for (int i = 2; i < n; i++) {
+        ll x; cin >> x; 
+        b = __gcd(b,x);
+        red();
+    }
+    cout << b << endl;
 }
 ```
 
+</CPPSection>
+
+<JavaSection>
+
+```java
+import java.util.*;
+import java.io.*;
+
+public class ocralcm {
+    public static void main(String[] args) throws IOException {
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+        int N = Integer.parseInt(br.readLine());
+        StringTokenizer st = new StringTokenizer(br.readLine());
+        int[] nums = new int[N];
+        
+        for(int i=0; i<N; i++){
+            nums[i] = Integer.parseInt(st.nextToken());
+        }
+
+        long gcd = gcd(nums[0], nums[1]);
+        long ret = nums[0]/gcd * nums[1];
+        for(int i = 2; i<N; gcd = gcd(gcd, nums[i]), i++){
+            ret = gcd(ret, lcm(nums[i], gcd));
+        }
+        System.out.println(ret);
+    }
+    
+    static long lcm(long a, long b){
+        return a/gcd(a,b) * b;
+    }      
+    
+    static long gcd(long a, long b){
+        return b == 0 ? a : gcd(b, a % b);
+    }
+}
+```
+
+</JavaSection>
+</LanguageSection>
+
 If it's hard to understand what exactly this code is doing at first glance, a good first step is to simulating the code in the case where each $a_i$ is a power of the same prime (say, $2^k$). If the algorithm works for this case, then it will also work in the general case since the contributions of different primes are computed independently and multiplied together.
 
 <!-- the GCD function is [multiplicative](https://en.wikipedia.org/wiki/Multiplicative_function). -->
+Here's another way of thinking about this:
+
+Consider fixing $j$. How can the gcd of all lcms of $a_j$, and all $a_i$ (such that $i$ is in range $[0,j]$) be quickly computed? The result would be $a_j*C$, where $C$ is the gcd of all $a_i$, and we can keep track of this gcd on the run.
```

Comment:
[0,j)?

Reply:
Wouldn't it be lcm(a_j,C)? Also, use `\cdot` instead of `*`

Reply:
Yes and yes. Not sure why I put a_j*C. Changes have been made.